vimeo-networking/src/main/java/com/vimeo/networking/utils/VimeoNetworkUtil.java

```diff
@@ -209,30 +172,62 @@ public void run() {
+    }
+
+    /**
+     * Creates a local API error with the provided error message and invalid parameters list.
+     *
+     * @param message           The message explaining the error.
+     * @param invalidParameters The parameters that were invalid.
+     */
+    @NotNull
+    public static VimeoResponse.Error.Api createLocalApiError(@NotNull final String message,
+                                                              @NotNull final InvalidParameter... invalidParameters) {
+        return new VimeoResponse.Error.Api(ApiErrorFactory.createApiErrorWithInvalidParameters(message,
+                                                                                               invalidParameters),
+                                           -1);
+    }
+
+    /**
+     * This utility method takes a Retrofit response and extracts a {@link ApiError} object out of it if
      * applicable. It will return null in the case where there has been a successful response.
      *
      * @param response A non-null response from the Vimeo API
-     * @return a {@link VimeoError} object extracted from the response or null
+     * @return a {@link VimeoResponse.Error} object extracted from the response or null
      */
-    @Nullable
-    public static <ResponseType_T> VimeoError getErrorFromResponse(@Nullable final Response<ResponseType_T> response) {
-        if (response != null && response.isSuccessful()) {
-            return null;
+    @NotNull
+    public static <ResponseType_T> VimeoResponse.Error getErrorFromResponse(@NotNull final Response<ResponseType_T> response) {
+        if (response.isSuccessful()) {
+            throw new IllegalStateException("Cannot get error from a successful response");
         }
-        VimeoError vimeoError = null;
-        if (response != null && response.errorBody() != null) {
+        final VimeoResponse.Error vimeoError;
+        String errorBody = null;
+        if (response.errorBody() != null) {
             try {
-                vimeoError = getGson().fromJson(response.errorBody().string(), VimeoError.class);
+                errorBody = response.errorBody().string();
+                final ApiError apiError = getMoshi().adapter(ApiError.class).fromJson(errorBody);
+                if (apiError != null) {
+                    return new VimeoResponse.Error.Api(apiError, response.code());
+                } else {
+                    return new VimeoResponse.Error.Unknown(errorBody, response.code());
+                }
             } catch (final Exception e) {
                 ClientLogger.e("Error while attempting to convert response body to VimeoError", e);
+                if (errorBody == null) {
+                    errorBody = "Unable to read error body";
+                }
+                return new VimeoResponse.Error.Unknown(errorBody, response.code());
             }
+        } else {
+            return new VimeoResponse.Error.Unknown("Null error body", response.code());
```

Comment:
Maybe "Null error body" and "Unable to read error body" could be pulled out into error constants so callers could potentially distinguish these errors? Or maybe we should pull these out as additional typed error classes? It seems weird to categorize them as "Unknown" types when they are very clearly known in advance.

Reply:
These errors are not expected to happen on a frequency which would justify us enumerating the cases in which they are created. Even when the Vimeo API returns a 500 it is supposed to return us a valid `ApiError`. The cases in which these `Unknown` errors will be received are exceptional and unless we see a large number of these errors, I don't want the class to be overengineered to handle scenarios that are unlikely to happen. 

These errors are "Unknown" in the sense that it is unknown _why_ they are happening. In the case of an `Api` or `Exception`, the reason for their occurrence is known, but in the case of an `Unknown`, it is unclear, for instance, why we would receive an unsuccessful response with a null error body, or why the error body would be non-null but unreadable. The purpose of `Unknown` is to act as a net for errors that don't fit into the other two categories. The reason we are not using `Exception` is that `Exception` is reserved for cases without an `httpStatusCode` (i.e. the server never responded).